projects/whydense/cifar/cifar_tune.py

```diff
@@ -0,0 +1,215 @@
+
+import argparse
+import time
+
+import ray
+from ray import tune
+from ray.tune.schedulers import *
+from torchvision import datasets
+
+from nupic.research.frameworks.pytorch.model_utils import *
+from nupic.research.frameworks.pytorch.image_transforms import *
+from nupic.research.frameworks.pytorch.cifar_experiment import TinyCIFAR
+
+import configparser
+
+
+class CIFARTune(TinyCIFAR, tune.Trainable):
+  """
+  ray.tune trainable class for running small CIFAR models:
+  - Override _setup to reset the experiment for each trial.
+  - Override _train to train and evaluate each epoch
+  - Override _save and _restore to serialize the model
+  """
+
+  def __init__(self, config=None, logger_creator=None):
+    TinyCIFAR.__init__(self)
+    tune.Trainable.__init__(self, config=config, logger_creator=logger_creator)
+
+
+  def _setup(self, config):
+    """Custom initialization.
+
+    Args:
+        config (dict): Hyperparameters and other configs given.
+            Copy of `self.config`.
+    """
+    self.model_setup(config)
+
+
+  def _train(self):
+    """Implement train() for a single epoch.
+
+    Returns:
+        A dict that describes training progress."""
+
+    return self.train_epoch(self._iteration)
+
+
+  def _save(self, checkpoint_dir):
+    return self.model_save(checkpoint_dir)
+
+
+  def _restore(self, checkpoint):
+    """Subclasses should override this to implement restore().
+
+    Args:
+        checkpoint (str | dict): Value as returned by `_save`.
+            If a string, then it is the checkpoint path.
+    """
+
+    self.model_restore(checkpoint)
+
+
+@ray.remote
+def run_experiment(config, trainable):
+  """
+  Run a single tune experiment in parallel as a "remote" function.
+
+  :param config: The experiment configuration
+  :type config: dict
+  :param trainable: tune.Trainable class with your experiment
+  :type trainable: :class:`ray.tune.Trainable`
+  """
+  # Stop criteria. Default to total number of iterations/epochs
+  stop_criteria = {
+    "training_iteration": config.get("iterations")
+  }
+  stop_criteria.update(config.get("stop", {}))
+
+  tune.run(
+    trainable,
+    name=config["name"],
+    local_dir=config["path"],
+    stop=stop_criteria,
+    config=config,
+    num_samples=config.get("repetitions", 1),
+    search_alg=config.get("search_alg", None),
+    scheduler=config.get("scheduler",
+                         MedianStoppingRule(
+                           time_attr="training_iteration",
+                           reward_attr='mean_accuracy',
+                           min_samples_required=3,
+                         )),
+    trial_executor=config.get("trial_executor", None),
+    checkpoint_at_end=config.get("checkpoint_at_end", False),
+    checkpoint_freq=config.get("checkpoint_freq", 0),
+    resume=config.get("resume", False),
+    reuse_actors=config.get("reuse_actors", False),
+    verbose=config.get("verbose", 0)
+  )
+
+def parse_config(config_file, experiments=None):
```

Comment:
Should use 'nupic.research.support.parse_config' instead

Reply:
I would like to, but the 'eval' didn't work for tune.grid_search. I think maybe it needs to be in the same scope as the parse_config function...??